nemo/collections/llm/gpt/data/pre_training.py

```diff
@@ -17,7 +17,8 @@
 class PreTrainingDataModule(pl.LightningDataModule):
```

Comment:
Could you split out an abstract pl.LightningDataModule that requires `self.data_sampler` to be set, and move the `state_dict` stuff into that? I'm guessing everyone who uses a `data_sampler` and wants to restore the step they were on will need this, right? I don't see anything specific to `weights` etc in those functions and I would like to inherit from a thing that's documented and not have to think about this. What do you think @marcromeyn ?

Reply:
Maybe it could live in nemo/lightning/data.py and then you could make `PreTrainingDataModule` inherit from that?

Reply:
Maybe something like this:
```
import pytorch_lightning as pl
from abc import ABC, abstractmethod

class NeMoDataModule(pl.DataModule, ABC):
    def __init__(self):
        super().__init__()
        self._data_sampler = None
        self.initialize_data_sampler()
        if self._data_sampler is None:
            raise ValueError("Child class must set self.data_sampler")

    @abstractmethod
    def initialize_data_sampler(self):
        """Method that must be implemented by the child class to set self.data_sampler."""
        pass

    @property
    def data_sampler(self):
        if self._data_sampler is None:
            raise ValueError("data_sampler must be set")
        return self._data_sampler

    @data_sampler.setter
    def data_sampler(self, value):
        self._data_sampler = value

class PreTrainingDataModule(NeMoDataModule):
    def __init__(self):
        super().__init__()
        # rest of setup, set state needed for later data_sampler initialization
    def initialize_data_sampler(self):
        # NOTE: this is called in the parent class during `super().__init__()`
        self.data_sampler = ... # initialize data sampler here 
```

Reply:
Makes sense, let me do that in a future PR